File: backend/app/routers/admin/bookings.py
from datetime import date, datetime

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings
from app.core.security import AdminClaims, get_current_admin, scope_property_id
from app.db import get_session
from app.models import Booking, BookingStatus
from app.schemas.admin_content import AdminBookingCreate
from app.services.booking import create_manual_booking

router = APIRouter(prefix="/api/admin/bookings", tags=["admin-bookings"])


@router.post("")
async def post_manual_booking(
    payload: AdminBookingCreate,
    admin: AdminClaims = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
):
    """Logs a walk-in/phone reservation taken outside the site — payment is
    handled by staff directly, so this creates the booking already confirmed."""
    property_id = scope_property_id(payload.property_id, admin)
    return await create_manual_booking(session, property_id, payload)


@router.get("")
async def list_bookings(
    property: int | None = None,
    status: BookingStatus | None = None,
    date_from: date | None = None,
    date_to: date | None = None,
    limit: int = Query(default=100, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    admin: AdminClaims = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
):
    scoped_property = scope_property_id(property, admin)
    query = select(Booking)
    if scoped_property is not None:
        query = query.where(Booking.property_id == scoped_property)
    if status is not None:
        query = query.where(Booking.status == status)
    if date_from is not None:
        query = query.where(Booking.check_in >= date_from)
    if date_to is not None:
        query = query.where(Booking.check_out <= date_to)
    result = await session.execute(query.order_by(Booking.created_at.desc()).limit(limit).offset(offset))
    return result.scalars().all()


async def _get_scoped_booking(session: AsyncSession, admin: AdminClaims, booking_id: int) -> Booking:
    booking = await session.get(Booking, booking_id)
    if not booking:
        raise HTTPException(404, "Booking not found")
    scoped = scope_property_id(booking.property_id, admin)
    if scoped != booking.property_id:
        raise HTTPException(404, "Booking not found")
    return booking


@router.patch("/{booking_id}/confirm")
async def confirm_booking(
    booking_id: int,
    admin: AdminClaims = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
):
    booking = await _get_scoped_booking(session, admin, booking_id)
    if booking.status != BookingStatus.pending_whatsapp:
        raise HTTPException(400, "Only pending WhatsApp bookings can be confirmed this way")
    booking.status = BookingStatus.confirmed
    booking.updated_at = datetime.utcnow()
    session.add(booking)
    await session.commit()
    await session.refresh(booking)
    return booking


@router.patch("/{booking_id}/cancel")
async def cancel_booking(
    booking_id: int,
    admin: AdminClaims = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
):
    booking = await _get_scoped_booking(session, admin, booking_id)
    booking.status = BookingStatus.cancelled
    booking.updated_at = datetime.utcnow()
    session.add(booking)
    await session.commit()
    await session.refresh(booking)
    return booking


@router.post("/{booking_id}/refund")
async def refund_booking(
    booking_id: int,
    admin: AdminClaims = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
):
    """Stub — flips status to refunded. Replace with a real Razorpay refund call
    once sandbox keys are wired up; payments.status must then only change via that call.
    Gated by payments_stub_mode so this can't fake a refund once a real gateway is live."""
    if not settings.payments_stub_mode:
        raise HTTPException(501, "Refund is not implemented")
    booking = await _get_scoped_booking(session, admin, booking_id)
    if booking.status != BookingStatus.confirmed:
        raise HTTPException(400, "Only confirmed bookings can be refunded")
    booking.status = BookingStatus.refunded
    booking.updated_at = datetime.utcnow()
    session.add(booking)
    await session.commit()
    await session.refresh(booking)
    return booking

File: backend/src/routes/admin/bookings.ts
import { Router } from "express";
import { queryOne, query } from "../../db";
import { asyncHandler, HttpError } from "../../middleware/errors";
import { validateBody } from "../../middleware/validate";
import { settings } from "../../config";
import { AdminBookingCreateSchema } from "../../schemas/adminContent";
import { AdminClaims, getCurrentAdmin, scopePropertyId } from "../../security";
import { createManualBooking } from "../../services/booking";
import { BookingStatus } from "../../types/enums";
import { Booking } from "../../types/models";

export const router = Router();

router.use(getCurrentAdmin);

router.post(
  "",
  validateBody(AdminBookingCreateSchema),
  asyncHandler(async (req, res) => {
    // Logs a walk-in/phone reservation taken outside the site — payment is
    // handled by staff directly, so this creates the booking already confirmed.
    const propertyId = scopePropertyId(req.body.property_id, req.admin!);
    const booking = await createManualBooking(propertyId, req.body);
    res.json(booking);
  })
);

router.get(
  "",
  asyncHandler(async (req, res) => {
    const admin = req.admin!;
    const scopedProperty = scopePropertyId(
      req.query.property !== undefined ? parseInt(String(req.query.property), 10) : null,
      admin
    );
    const status = typeof req.query.status === "string" ? req.query.status : undefined;
    const dateFrom = typeof req.query.date_from === "string" ? req.query.date_from : undefined;
    const dateTo = typeof req.query.date_to === "string" ? req.query.date_to : undefined;
    const limit = Math.min(Math.max(parseInt(String(req.query.limit ?? "100"), 10) || 100, 1), 200);
    const offset = Math.max(parseInt(String(req.query.offset ?? "0"), 10) || 0, 0);

    const conditions: string[] = [];
    const params: unknown[] = [];
    if (scopedProperty !== null) {
      params.push(scopedProperty);
      conditions.push(`property_id = $${params.length}`);
    }
    if (status) {
      params.push(status);
      conditions.push(`status = $${params.length}`);
    }
    if (dateFrom) {
      params.push(dateFrom);
      conditions.push(`check_in >= $${params.length}`);
    }
    if (dateTo) {
      params.push(dateTo);
      conditions.push(`check_out <= $${params.length}`);
    }
    const where = conditions.length ? `WHERE ${conditions.join(" AND ")}` : "";
    params.push(limit, offset);
    const rows = await query<Booking>(
      `SELECT * FROM bookings ${where} ORDER BY created_at DESC LIMIT $${params.length - 1} OFFSET $${params.length}`,
      params
    );
    res.json(rows);
  })
);

async function getScopedBooking(admin: AdminClaims, bookingId: number): Promise<Booking> {
  const booking = await queryOne<Booking>("SELECT * FROM bookings WHERE id = $1", [bookingId]);
  if (!booking) throw new HttpError(404, "Booking not found");
  const scoped = scopePropertyId(booking.property_id, admin);
  if (scoped !== booking.property_id) throw new HttpError(404, "Booking not found");
  return booking;
}

router.patch(
  "/:bookingId/confirm",
  asyncHandler(async (req, res) => {
    const booking = await getScopedBooking(req.admin!, parseInt(req.params.bookingId, 10));
    if (booking.status !== BookingStatus.pending_whatsapp) {
      throw new HttpError(400, "Only pending WhatsApp bookings can be confirmed this way");
    }
    const updated = await queryOne<Booking>(
      `UPDATE bookings SET status = $1, updated_at = now() WHERE id = $2 RETURNING *`,
      [BookingStatus.confirmed, booking.id]
    );
    res.json(updated);
  })
);

router.patch(
  "/:bookingId/cancel",
  asyncHandler(async (req, res) => {
    const booking = await getScopedBooking(req.admin!, parseInt(req.params.bookingId, 10));
    const updated = await queryOne<Booking>(
      `UPDATE bookings SET status = $1, updated_at = now() WHERE id = $2 RETURNING *`,
      [BookingStatus.cancelled, booking.id]
    );
    res.json(updated);
  })
);

router.post(
  "/:bookingId/refund",
  asyncHandler(async (req, res) => {
    // Stub — flips status to refunded. Replace with a real Razorpay refund call
    // once sandbox keys are wired up; payments.status must then only change via that call.
    // Gated by payments_stub_mode so this can't fake a refund once a real gateway is live.
    if (!settings.paymentsStubMode) throw new HttpError(501, "Refund is not implemented");
    const booking = await getScopedBooking(req.admin!, parseInt(req.params.bookingId, 10));
    if (booking.status !== BookingStatus.confirmed) throw new HttpError(400, "Only confirmed bookings can be refunded");
    const updated = await queryOne<Booking>(
      `UPDATE bookings SET status = $1, updated_at = now() WHERE id = $2 RETURNING *`,
      [BookingStatus.refunded, booking.id]
    );
    res.json(updated);
  })
);
